fabric-mining-levels-v0/src/main/java/net/fabricmc/fabric/api/tools/DynamicTool.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tools;
+
+import net.minecraft.item.ItemStack;
+
+/**
+ * Interface for tools whose properties are affected by NBT. Goes on the item.
+ */
+public interface DynamicTool {
+	/**
+	 * @param stack The stack to check on.
+	 * @return The mining level of the tool. 3 is equal to a diamond pick.
+	 */
+	int getMiningLevel(ItemStack stack);
+
+	/**
+	 * @param stack The stack to check on.
+	 * @return The mining speed of the tool. 8.0 is equal to a diamond pick.
+	 */
+	float getMiningSpeed(ItemStack stack);
```

Comment:
Hm, we might wanna do some rework on how the whole system is set up. Currently we basically skip vanilla behavior for anything that's an instanceof `MiningToolItem` so that we can use our own, non-hardcoded `breakByTool` system, but we might wanna just inject into `MiningToolItem` and change logic there instead. I'll bring it up in the discord and see what people think.

Reply:
Looking at it, the `Item#getMiningSpeed` is what handles speed for effectiveness too, so it's not super cut out for our purposes. The system is currently designed so `ToolManager` is supposed to be internal-only, so we shouldn't set up a scenario where devs have to call ToolManager for effectiveness. There's no way to make sure that all subclasses of MiningToolItem will call to super isEffectiveOn if we use the code there, so it'd be a hassle. We currently implement this stuff in a mixin to ItemStack, which I think is the best option. I'm gonna say we keep the getMiningSpeed option in.